crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -598,3 +509,146 @@ impl TryFrom<TransactionBlockEffects> for TransactionBlock {
     }
 }
+
+fn apply_scan_limited_pagination(
+    conn: &mut ScanConnection<String, TransactionBlock>,
+    page: &Page<Cursor>,
+    transactions: &Vec<StoredTransaction>,
+    tx_bounds: TxBounds,
+    scan_limit: Option<u64>,
+    checkpoint_viewed_at: u64,
+) {
+    if page.is_from_front() {
+        apply_forward_scan_limited_pagination(
+            conn,
+            page,
+            transactions,
+            tx_bounds,
+            scan_limit,
+            checkpoint_viewed_at,
+        );
+    } else {
+        apply_backward_scan_limited_pagination(
+            conn,
+            page,
+            transactions,
+            tx_bounds,
+            scan_limit,
+            checkpoint_viewed_at,
+        );
+    }
+}
+
+/// In the forward scan-limited pagination scenario, there are two special cases when dealing with
+/// the new page's `startCursor`. If given a scan-limited cursor, `paginate_results` will expect to
+/// find this cursor in the results, and subsequently report that there is no preceding or next page
+/// because the cursor would not actually show up. In another scenario, if the cursor is not from
+/// scan limit, the cursor will be found in the results and pruned. If the resulting set is empty,
+/// `paginate_results` will still report that there is no previous and next page for it. However, if
+/// a `scan_limit` has been set, we want to override this. In these two scenarios, we would set the
+/// new `startCursor` to one past the previous cursor.
+///
+/// When paginating forward, if `paginate_results` indicates that there is no next page and we have
+/// a scan limit, this might be a false end, because we haven't scanned all transactions. Then we
+/// indicate there is a next page if the scan limit does not reach the end of the tx bounds. The
+/// actual cursor values depend on a few things. If the result set is less than the `limit` and the
+/// page flag is false, including an empty set, then we know we have exhausted the candidates to
+/// scan, and we can set the ending cursor to the last transaction scanned. Otherwise, we'll use the
+/// last element's cursor.
+fn apply_forward_scan_limited_pagination(
+    conn: &mut ScanConnection<String, TransactionBlock>,
+    page: &Page<Cursor>,
+    transactions: &Vec<StoredTransaction>,
+    tx_bounds: TxBounds,
+    scan_limit: Option<u64>,
+    checkpoint_viewed_at: u64,
+) {
+    if let Some(after) = page.after() {
+        if (after.is_scan_limited || scan_limit.is_some() && transactions.is_empty())
+            && after.tx_sequence_number > tx_bounds.lo
+        {
+            let mut start_cursor = cursor::TransactionBlockCursor {
+                checkpoint_viewed_at,
+                tx_sequence_number: after.tx_sequence_number + 1,
+                is_scan_limited: true,
+            };
+
+            start_cursor.is_scan_limited = transactions.is_empty()
+                || transactions[0].tx_sequence_number as u64 != start_cursor.tx_sequence_number;
```

Comment:
I think this should always be true if we are setting the cursor (which we seem to be doing unconditionally), because the cursor we are picking is not guaranteed to be a valid result for the query. This is true even if `after` was found in the results and was pruned.

Reply:
ah yes ... you're right. by the time it gets here, the first element will never match the cursor

Reply:
Oh, I recall now ... this is so that when we return results so that `results.len() < limit`, we set the `endCursor` to the last transaction scanned and `isScanLimited=true`

Reply:
but i think the conditional can be clearer or i can add some comments here

Reply:
ok i think we need to keep this, and updated the comments to reflect

In the forward pagination scenario, to determine the correct `startCursor` for a scan-limited page:

1. `after.is_some() && after.is_scan_limited()`
2. or if `scan_limit.is_some() && after.is_some()`

then we want to adjust the new page's `startCursor` to `after.tx_sequence_number + 1`. We then check if this cursor value is at the tip of the `transactions` array, and if so, then we should set `is_scan_limited` to false, otherwise true by default

to determine the correct `endCursor`
- only if `!conn.has_next_page and scan_limit.is_some() && tx_bounds.scan_has_next_page()`
There are 4 scenarios that will yield `has_next_page=false`:
1. met `limit`, `paginate_results` doesn't detect more results + more to scan
2. met `limit`, `paginate_Results` doesn't detect more results + no more to scan
3. less than `limit`, `paginate_results` doesn't detect more results + more to scan
4. less than `limit`, `paginate_results` doesn't detect more results + no more to scan
Regardless of the scenario, we can set the `endCursor` to the last transaction scanned.